src/hssm/param/params.py

```diff
@@ -0,0 +1,353 @@
+            raise ValueError("Parent parameter not found.")
+        result_formula: bmb.Formula = bmb.Formula(parent_formula, *other_formulas)
+        result_priors = None if not priors else priors
+        result_links: dict | str = "identity" if not links else links
+
+        return result_formula, result_priors, result_links
+
+    def __str__(self) -> str:
+        """Return the representation of the class.
+
+        Returns
+        -------
+        str
+            A string representation of the class.
+        """
+        return self.__repr__()
+
+    def __repr__(self):
+        """Return the representation of the class.
+
+        Returns
+        -------
+        str
+            A string representation of the class.
+        """
+        params = "\n".join(repr(param) for _, param in self.items())
+        return "Parameters:\n" f"{params}"
+
+
+def collect_user_params(
+    model: HSSM,
+    include: list[dict[str, Any] | UserParam],
+    kwargs: dict[str, Any],
+    p_outlier: float | dict | bmb.Prior | None,
+) -> dict[str, UserParam]:
+    """Collect and convert parameters to UserParam objects.
+
+    Parameters
+    ----------
+    model
+        The HSSM model.
+    include
+        A list of dictionaries or UserParam objects specifying the parameters.
+    kwargs
+        Keyword arguments specifying the parameters.
+    p_outlier
+        The prior specification for the outlier probability.
+
+    Returns
+    -------
+    dict[str, UserParam]
+        A dictionary with the parameter names as keys and UserParam objects as values.
+    """
+    user_params: dict[str, UserParam] = {}
+
+    # Process include
+    for param in include:
+        user_param = UserParam.from_dict(param) if isinstance(param, dict) else param
+        if user_param.name is None:
+            raise ValueError("Parameter name must be specified.")
+        if user_param.name not in model.list_params:
+            raise ValueError(f"Parameter {user_param.name} not found in list_params.")
+        if user_param.name == "p_outlier":
+            raise ValueError(
+                "Please do not specify `p_outlier` in `include`. "
+                "Please specify it with `p_outlier` instead."
+            )
+        user_params[user_param.name] = user_param
+
+    # Process kwargs
+    # If any of the keys is found in `list_params` it is a parameter specification
+    # We add the parameter specification to `user_params` and remove it from
+    # `kwargs`
+    for param_name in model.list_params:
+        if param_name in kwargs:
+            if param_name in user_params:
+                raise ValueError(
+                    f"Parameter `{param_name}` specified in both"
+                    " `include` and `kwargs`."
+                )
+            user_params[param_name] = UserParam.from_kwargs(
+                param_name,
+                # Use pop to remove the key from kwargs
+                kwargs.pop(param_name),
+            )
```

Comment:
Consider a guard clause:
```suggestion
        if param_name in user_params and param_name in kwargs:
            raise ValueError(
                f"Parameter {param_name} specified in both "
                "include and kwargs."
            )

        # Update user_params only if param_name is in kwargs
        if param_name in kwargs:
            user_params[param_name] = UserParam.from_kwargs(
                param_name,
                kwargs.pop(param_name),
            )
```

Reply:
This way `param_name in kwargs` is checked twice right?

Reply:
Yes, but it avoids the nested if statement. One could save the check to a variable but that would add an extra line of code anyway. If `param_name` is always in `kwargs` then the update could be done without the check.

Reply:
I think I am OK with just one layer of nesting if that avoids one unnecessary check :). Is the readability that different?

Reply:
To me it is. What about a compromise saving the check to a variable: `param_in_user_params = param_name in user_params` and use it in the second if block?

Reply:
Okay. I think I'll compromise without adding this other variable. But in the future maybe we can agree to disagree and not making this as an issue in code reviews?

Reply:
Surely. Do note that I already approved the PR!